### PR TITLE
feat: allow `upgrade` command to accept multiple packages as arguments

### LINKED ISSUE #1336: Allow `pipx upgrade` to accept multiple packages as arguments
**How would this feature be useful?**

The possibility to handle multiple package specs at once was already introduced to the `install` command in #1102. The logical next step would be to extend this functionality to the `pipx upgrade`, and probably also `pipx reinstall` commands.

**Describe the solution you'd like**

Implement it.

**Describe alternatives you've considered**

Alternatively, one could write shell script or other programming code wrappers to take over this functionality, which however is not exactly user-friendly nor efficient.

### PR DESCRIPTION
**Pull request by huxuan**
- [x] I have added a news fragment under `changelog.d/` (if the patch affects the end users)

## Summary of changes

Close #1336

## Test plan

Tested by running

```
# command(s) to exercise these changes
nox -s tests
```

### CODE REVIEW

File: src/pipx/commands/upgrade.py [hunk 1]
```diff
@@ -192,17 +192,20 @@ def upgrade(
...
+            pip_args,
+            verbose,
+            include_injected=include_injected,
+            upgrading_all=False,
+            force=force,
+            install=install,
+            python=python,
+            python_flag_passed=python_flag_passed,
+        )
+        for venv_dir in venv_dirs.values()
```
[the remark this answers is not in the document]
Comment (huxuan) [reply src/pipx/commands/upgrade.py]: Done.

File: src/pipx/main.py [hunk 2]
```diff
@@ -195,6 +195,14 @@ def run_pipx_command(args: argparse.Namespace, subparsers: Dict[str, argparse.Ar
 
         venv_dir = venv_container.get_venv_dir(package)
         logger.info(f"Virtual Environment location is {venv_dir}")
+
+    if "packages" in args:
+        if any(urllib.parse.urlparse(package).scheme for package in args.packages):
+            raise PipxError("Package cannot be a url")
```
Comment (dukecat0) [inline src/pipx/main.py:201]:
```suggestion
            raise PipxError("Package cannot be a url. Package name should be passed instead.")
```

File: tests/test_upgrade.py [hunk 3]
```diff
@@ -102,3 +102,17 @@ def test_upgrade_install_missing(pipx_temp_env, capsys):
     assert not run_pipx_cli(["upgrade", "pycowsay", "--install"])
     captured = capsys.readouterr()
     assert "installed package pycowsay" in captured.out
+
+
+def test_upgrade_multiple(pipx_temp_env, capsys):
+    name = "pylint"
+    pkg_spec = PKG[name]["spec"]
+    initial_version = pkg_spec.split("==")[-1]
+    assert not run_pipx_cli(["install", f"{pkg_spec}"])
```
Comment (dukecat0) [inline tests/test_upgrade.py:111]: Looks like the f-string isn't necessary here?
```suggestion
    assert not run_pipx_cli(["install", pkg_spec])
```

File: tests/test_upgrade.py [hunk 4]
```diff
@@ -102,3 +102,17 @@ def test_upgrade_install_missing(pipx_temp_env, capsys):
...
+
+    assert not run_pipx_cli(["install", "pycowsay"])
+
+    assert not run_pipx_cli(["upgrade", f"{name}", "pycowsay"])
```
Comment (dukecat0) [inline tests/test_upgrade.py:115]:
```suggestion
    assert not run_pipx_cli(["upgrade", name, "pycowsay"])
```

**Review by dukecat0 — APPROVED** [review]

**Comment by huxuan** [comment]
Hi @dukecat0, 

Sorry, I just copied some code snippets from the surrounding test cases, I should be more careful.

Thanks for pointing out.

**Review by uranusjr — APPROVED** [review]

File: src/pipx/main.py [hunk 5]
```diff
@@ -195,6 +195,14 @@ def run_pipx_command(args: argparse.Namespace, subparsers: Dict[str, argparse.Ar
 
         venv_dir = venv_container.get_venv_dir(package)
         logger.info(f"Virtual Environment location is {venv_dir}")
+
+    if "packages" in args:
+        if any(urllib.parse.urlparse(package).scheme for package in args.packages):
```
Comment (chrysle) [inline src/pipx/main.py:200]: BTW, this doesn't validate whether a package is a URL correctly. It should at least also check whether it contains a network location. The current form would detect a Windows path as a URL. See the [docs](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlparse).

Comment (huxuan) [reply src/pipx/main.py:200]: Hi @chrysle, any suggestions to check it more precisely?

Comment (chrysle) [reply src/pipx/main.py:200]: I've encountered this just recently and fixed it in https://github.com/pypa/pipx/pull/1329/commits/e94c5d2a92338012a12067ba049182d72aaa1f35.

Comment (huxuan) [reply src/pipx/main.py:200]: Hi @chrysle, seems the commit is not merged yet, should I send a pull request for this or for the `packages` part only?

Comment (huxuan) [reply src/pipx/main.py:200]: Just a reminder, seems the #1329 is merged. Maybe I should also make the same change to packages?

Comment (chrysle) [reply src/pipx/main.py:200]: Let's do it.

**Pull request merged**